### DISCUSSION TITLE
Understanding file-paths / storage usage

### DISCUSSION THREAD

**Discussion by andrewlow** [photoprism/photoprism#956] [Q&A] [answered]
In my setup (pure docker) I've got the following volume maps

```
                -v /media/archive/photoprism/storage:/photoprism/storage \
                -v /media/archive/photoprism/originals:/photoprism/originals \
                -v /media/archive/photoprism/import:/photoprism/import \
```

The `originals` path I believe is the tree of all photos as imported into the system.
`import` is a temporary staging area for new uploaded photos etc (I'm a fan of 'Move'-ing the files from import into originals)

Now 'storage' is a bunch of meta-data about your library. Stuff like jpg versions of your raw files etc, to make photoprism work better.

All good so far.

I'm also a user of `snapraid` for managing my media storage. Photos count as media. Clearly I want my `originals` folder to be managed by snapraid.  One of the limitations of snapraid is that it's not suitable for high turn-over files. So things like my Plex library of large video files ripped from DVDs, snapraid is amazing. I get reliable backups with checks for bitrot etc. 

My nightly scripts are showing me unfortunately high turn over for a bunch of 'yml' files managed by photoprism - even when the library is idle all day.

```
update photoprism/storage/albums/folder/aqmqiw03h8032yrg.yml
update photoprism/storage/albums/folder/aqmqj2hn3drxxofj.yml
update photoprism/storage/albums/folder/aqmqj461vg9yydlu.yml
update photoprism/storage/albums/folder/aqmqj8e2pb8qgcbs.yml
update photoprism/storage/albums/month/aqn9j9q3vzrvtdsc.yml
update photoprism/storage/albums/month/aqn9j9q1d027ka7m.yml
update photoprism/storage/albums/month/aqn9j9qwjy2f12xz.yml
update photoprism/storage/albums/month/aqn9j9qr24ephzza.yml
update photoprism/storage/albums/month/aqn9j9q2kjk0qwnu.yml
update photoprism/storage/albums/state/aqmqkb714a5fdbsn.yml
update photoprism/storage/albums/state/aqmqkb72gdqj184b.yml
update photoprism/storage/albums/state/aqmqkb72kgoe60r9.yml
update photoprism/storage/albums/state/aqmqkb75ih8n3lhi.yml
update photoprism/storage/albums/state/aqmqkb7h8hynwosn.yml
```

Now, `snapraid` also doesn't really like lots of little files - but I'm willing to live with the pain I'm causing myself with all the thumbnails etc being managed sub-optimally there. However, I really can't deal with 100's of file changes every day when nothing is happening to the library.

Are all of these yml files because I'm getting the default SQLite database? https://docs.photoprism.org/getting-started/faq/#why-is-my-storage-folder-so-large-what-is-in-it

I do have a few options - I could add an exclude rule to snapraid to ignore these tree's of files - but if there is a re-configuration (like running with a db) I could do to avoid this churn, that'd be even better. 

I'm also concerned that the yml files will be generated as part of the backup process https://docs.photoprism.org/getting-started/faq/#yaml - unless I intentionally disable them.

Again, awesome software - thanks for building it. Hopefully this question is clear - and helps others understand things better once we arrive at an answer.

**Comment by alexisLefebvre** [comment]
Storage is “Permanent storage for settings, index & sidecar files (DON'T REMOVE):”.

Source: https://github.com/photoprism/photoprism/blob/0ffdc9c457508df42cdde0d6bc64ec22c3be843e/docker/examples/docker-compose.yml#L84-L85

---

In `storage`, we have these directories:
- `albums`
- `cache`
- `settings`
- `sidecar`

If you want to save space, you could ignore the `cache` and `sidecar` directories in your backup because these files can be generated again later even if you lose them.

**Reply by andrewlow** [reply]
Thanks for the reply, but you seem to have missed the point that I'm seeing regularly changing files in `storage/albums` - one of the tree's that you've indicated should be backed up..

Also - I'd caution others - as the suggestion, also means that the index.db created by SQLite would land inside of the docker container filesystem - causing it to be lost in each update. (but I also didn't test this, so maybe I'm wrong)

**Reply by alexisLefebvre** [reply]
I'm sorry I don't know why these files are updated.

---

You're right, I removed the example configs with several paths in the `storage/` directory. I don't know Docker enough to show a safe config.

**Reply by alexisLefebvre** [reply]
Disabling the YAML backup may remove these updates (see lower right):

> ![image](https://user-images.githubusercontent.com/2071331/105755439-9613d480-5f4b-11eb-9660-25cba12635c7.png)

**Reply by andrewlow** [reply]
I suspect you are correct as I linked in my original post: https://docs.photoprism.org/getting-started/faq/#yaml

I like backups :) I just want to minimize file churn for a static library.

Still, it may be worth exploring to see if disabling backups of to YAML will stop the zillion files from changing (then I just deal with the index changing often). 

My ideal here - would be for photoprism to allow me to put frequently changing files - in a unique path.  However it this is purely the backups happening - if I switch to a real database AND disable backups to YAML, I could manage backing up the db from time to time.

**Comment by andrewlow** [comment]
Oh look - `PHOTOPRISM_BACKUP_PATH` from https://docs.photoprism.org/getting-started/config-options/

This might be the answer. I think I can configure WHERE the backup lands - and this will allow me to send the noisy YAML files to a path where I don't have any `snapraid` conflicts.

I'll have to try this out.

**Reply by alexisLefebvre** [reply]
I think it's used when creating backup of the database by using CLI: https://github.com/photoprism/photoprism/search?q=PHOTOPRISM_BACKUP_PATH

**Reply by andrewlow** [reply]
I should have searched the source :)

Yeah - that looks like something different.

**Comment by graciousgrey** [comment, marked as answer]
We added more detailed docs regarding the folders used by PhotoPrism (https://docs.photoprism.org/user-guide/advanced/folders/) and backups (https://docs.photoprism.org/user-guide/advanced/backups/)

**Reply by andrewlow** [reply]
👍 

Thank you for updating the documentation - this is much clearer.

As I've gained more experience operating my photoprism install - I've landed here:

- Use MariaDB, this means 2 containers vs. 1 but is simply a more mainstream path and faster (use docker-compose to keep it all simple)
- If the YAML files bother you, `PHOTOPRISM_DISABLE_BACKUPS="true"` will stop them from being generated
- For ever file uploaded - many thumbnail / meta-data files are created anyways. `/photoprism/storage` is going to be noisy 

The great news is that `/photoprism/originals` is always very nicely 1:1 to your photos - meaning it's a good file tree to consider for additional backups/archiving.

With the additional documentation - it seems even within the storage tree there is some control you can have over the set of files being created there. This is great if you want this fine grained control, but as I suspect the maintainers already know - most users will not want to get into those details.
